### PR TITLE
Add DINOv2 depth estimation

### LINKED ISSUE #26057: Dinov2 for depth estimation
### Feature request

Dinov2's original repo has an example using Dinov2 backbone + DPT head for depth estimation [notebook link](https://github.com/facebookresearch/dinov2/blob/main/notebooks/depth_estimation.ipynb). If we can integrate it into `transformers` repo by adding a class `Dinov2ForImageDepthEstimation` and let `forward` method return `DepthEstimatorOutput`, we'll have a unified output interface across all depth estimation models. By doing this, we can easily chain this powerful depth estimation method together with other models under `transformers`'s pipelines. 

### Motivation

This would be a very great feature for many production use cases or research problems. One example is camera angle estimation from a 2D image, in which reliable depth information are critical. In my limited test cases, using dinov2+DPT head to run depth estimation is way better than the existing [DPT model](https://huggingface.co/docs/transformers/main/model_doc/dpt) itself. 

### Your contribution

I can submit a PR to add this feature if other professional developers don't have the bandwidth to deal with it. (I am relatively new to `transformers`'s develop workflow though.)

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

Fixes #26057 

PR that implements a part of #25799. It extends the DPT framework to use the `AutoBackbone` class. Next it uses `Dinov2Backbone` to convert the DINOv2+DPT checkpoints released by the authors [here](https://github.com/facebookresearch/dinov2/tree/main#pretrained-heads---depth-estimation).

To do:

- [x] add test in common backbone tests file to verify `out_indices` are saved properly => done in #26094 
- [x] add support for transforms in `DPTImageProcessor`
- [x] add tests which test DPT with a backbone
- [x] add integration test and convert remaining checkpoints

### CODE REVIEW

File: src/transformers/models/dpt/modeling_dpt.py [hunk 1]
```diff
@@ -716,14 +732,15 @@ def __init__(self, config):
...
         self.use_batch_norm = config.use_batch_norm_in_fusion_residual
+
         self.activation1 = ACT2FN["relu"]
         self.convolution1 = nn.Conv2d(
             config.fusion_hidden_size,
             config.fusion_hidden_size,
             kernel_size=3,
             stride=1,
             padding=1,
-            bias=not self.use_batch_norm,
```
Comment (amyeroberts) [inline src/transformers/models/dpt/modeling_dpt.py:726]: This isn't backwards compatible, users might have custom configs with non-default setting for `use_batch_norm_in_fusion_residual`. We need to do something along the lines of the following: 

* Default value of `use_bias_in_fusion_residual` and `use_batch_norm_in_fusion_residual` are `None` in the config
* If `use_batch_norm_in_fusion_residual` is set, then `use_bias_in_fusion_residual` takes this value, we emit a warning that this will change in the future and they should use `use_bias_in_fusion_residual`
* If both `use_batch_norm_in_fusion_residual` and `use_bias_in_fusion_residual` are set, then we use `use_bias_in_fusion_residual` and emit a warning that `use_batch_norm_in_fusion_residual` is being ignored
* If neither are set then `use_bias_in_fusion_residual` takes the default value of `True`

File: tests/models/dinov2/test_modeling_dinov2.py [hunk 2]
```diff
@@ -168,7 +168,7 @@ def create_and_check_backbone(self, config, pixel_values, labels):
         # verify feature maps
         self.parent.assertEqual(len(result.feature_maps), 1)
         self.parent.assertListEqual(
-            list(result.feature_maps[0].shape), [self.batch_size, self.seq_length, self.hidden_size]
+            list(result.feature_maps[0].shape), [self.batch_size, self.seq_length - 1, self.hidden_size]
```
Comment (amyeroberts) [inline tests/models/dinov2/test_modeling_dinov2.py:171]: Why did this change? This is possibly breaking for people using the backbone

File: src/transformers/models/dpt/modeling_dpt.py [hunk 3]
```diff
@@ -1297,7 +1339,7 @@ def forward(
 
             hidden_states = backbone_hidden_states
 
-        hidden_states = self.neck(hidden_states)
+        hidden_states = self.neck(hidden_states, None)
```
Comment (amyeroberts) [inline src/transformers/models/dpt/modeling_dpt.py:1342]: Let's pass these in a keyword arguments so it's clearer what each value represents

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this! 

Main comment is about backwards compatibility with config arguments. Otherwise looks good and just a few small nits. 

cc @rafaelpadilla

File: src/transformers/models/dpt/modeling_dpt.py [hunk 4]
```diff
@@ -634,12 +642,16 @@ def forward(self, hidden_states: List[torch.Tensor]) -> List[torch.Tensor]:
 
         for i, hidden_state in enumerate(hidden_states):
             if i not in self.neck_ignore_stages:
-                # reshape to (B, C, H, W)
-                hidden_state, cls_token = hidden_state[:, 1:], hidden_state[:, 0]
-                batch_size, sequence_length, num_channels = hidden_state.shape
-                size = int(math.sqrt(sequence_length))
-                hidden_state = hidden_state.reshape(batch_size, size, size, num_channels)
-                hidden_state = hidden_state.permute(0, 3, 1, 2).contiguous()
+                if hidden_state.ndim == 3:
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/dpt/modeling_dpt.py:645]: Then it is assumed to be a tensor of shape (batch_size, num_channels, height), width), representing a 4D feature map from the backbone.

File: src/transformers/models/dpt/modeling_dpt.py [hunk 5]
```diff
@@ -663,7 +675,11 @@ class DPTReassembleLayer(nn.Module):
     def __init__(self, config, channels, factor):
         super().__init__()
         # projection
-        self.projection = nn.Conv2d(in_channels=config.hidden_size, out_channels=channels, kernel_size=1)
+        if config.backbone_config is not None and config.is_hybrid is False:
+            hidden_size = config.backbone_config.hidden_size
+        else:
+            hidden_size = config.hidden_size
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/dpt/modeling_dpt.py:681]: Do you mean added to the `PretrainedModel` class?

[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/models/dpt/modeling_dpt.py:681]: I meant as a function, rather than method, but either probably works

File: src/transformers/models/dpt/modeling_dpt.py (hunk 1 above)
Comment (NielsRogge) [reply src/transformers/models/dpt/modeling_dpt.py:726]: Agree on keeping backwards compatibility, however I think this will throw warnings for each case, which is something we might need to avoid (I'm in favor of trying to avoid warnings as much as possible for user experience).

With the approach above, we cause the more popular [DPT-midas](https://huggingface.co/Intel/dpt-hybrid-midas) (> 2 million downloads) and [DPT-large](https://huggingface.co/Intel/dpt-large) (nearly 300k downloads) to always have a warning, whereas this batch norm was only used for [DPT-ade](https://huggingface.co/Intel/dpt-large-ade) (which almost no one is using - it has 800 downloads).

Hence I've pushed a change which tries to avoid warnings, let me know what you think

File: tests/models/dinov2/test_modeling_dinov2.py (hunk 2 above)
Comment (NielsRogge) [reply tests/models/dinov2/test_modeling_dinov2.py:171]: Yes it's a breaking change. However  the backbone is currently not part of the docs, has been added very recently, so I hope you're ok with this change. It now reflects the original implementation, which was not the case before.

File: src/transformers/models/dinov2/modeling_dinov2.py [hunk 6]
```diff
@@ -63,6 +64,35 @@
...
+@dataclass
+class Dinov2BackboneOutput(BackboneOutput):
+    """
+    Class for the outputs of [`Dinov2Backbone`].
+
+    Args:
+        feature_maps (`tuple(torch.FloatTensor)` of shape `(batch_size, num_channels, height, width)` or `(batch_size, sequence_length, hidden_size)`):
+            Feature maps of the stages. The shape depends on `config.reshape_hidden_states`.
+        cls_tokens (`tuple(torch.FloatTensor)` of shape `(batch_size, hidden_size)`):
+            Features of the CLS tokens of the stages.
```
Comment (ArthurZucker) [inline src/transformers/models/dinov2/modeling_dinov2.py:76]:
```suggestion
            Features of the CLS tokens for each stage.
```

File: src/transformers/models/dinov2/modeling_dinov2.py [hunk 7]
```diff
@@ -63,6 +64,35 @@
...
 
 
+@dataclass
+class Dinov2BackboneOutput(BackboneOutput):
+    """
+    Class for the outputs of [`Dinov2Backbone`].
+
+    Args:
+        feature_maps (`tuple(torch.FloatTensor)` of shape `(batch_size, num_channels, height, width)` or `(batch_size, sequence_length, hidden_size)`):
+            Feature maps of the stages. The shape depends on `config.reshape_hidden_states`.
```
Comment (ArthurZucker) [inline src/transformers/models/dinov2/modeling_dinov2.py:74]:
```suggestion
            Feature maps of each stage. The shape depends on `config.reshape_hidden_states`.
```

File: src/transformers/models/dinov2/modeling_dinov2.py [hunk 8]
```diff
@@ -858,8 +892,9 @@ def forward(
                 output = (feature_maps,) + outputs[2:]
             return output
 
-        return BackboneOutput(
+        return Dinov2BackboneOutput(
```
Comment (ArthurZucker) [inline src/transformers/models/dinov2/modeling_dinov2.py:895]: this is kind of breaking if the tuples have the same behaviour, and not very consistent in case `cls_tokens` can only be return if `return_dict = True`. But better than breaking 😉

File: src/transformers/models/dinov2/modeling_dinov2.py [hunk 9]
```diff
@@ -858,8 +892,9 @@ def forward(
                 output = (feature_maps,) + outputs[2:]
             return output
```
Comment (ArthurZucker) [inline src/transformers/models/dinov2/modeling_dinov2.py:893]: if not return dict should also return the `cls_tokens`. Unless in order not to be breaking you have to set `return_dict` to `True` to get the `cls_tokens`?

File: src/transformers/models/dpt/convert_dinov2_depth_to_hf.py [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+    # Verify forward pass
+    with torch.no_grad():
+        outputs = model(pixel_values)
+
+    predicted_depth = outputs.predicted_depth
+
+    print("Shape of predicted depth:", predicted_depth.shape)
+    print("First values of predicted depth:", predicted_depth[0, :3, :3])
+
+    # assert logits
```
Comment (ArthurZucker) [inline src/transformers/models/dpt/convert_dinov2_depth_to_hf.py:300]: same comment about testing the logits. This should not be on by default on should be done outside if really necessary. Otherwise a test on the test_modeling should suffice.

File: src/transformers/models/dpt/image_processing_dpt.py [hunk 11]
```diff
@@ -208,6 +218,51 @@ def resize(
             **kwargs,
         )
 
+    def pad_image(
```
Comment (ArthurZucker) [inline src/transformers/models/dpt/image_processing_dpt.py:221]: was this copied from?

File: src/transformers/models/dpt/modeling_dpt.py [hunk 12]
```diff
@@ -619,12 +626,13 @@ def _init_reassemble_dpt(self, config):
...
             self.readout_projects = nn.ModuleList()
+            hidden_size = self.get_backbone_hidden_size()
             for _ in range(len(config.neck_hidden_sizes)):
                 self.readout_projects.append(
-                    nn.Sequential(nn.Linear(2 * config.hidden_size, config.hidden_size), ACT2FN[config.hidden_act])
+                    nn.Sequential(nn.Linear(2 * hidden_size, hidden_size), ACT2FN[config.hidden_act])
                 )
 
-    def forward(self, hidden_states: List[torch.Tensor]) -> List[torch.Tensor]:
+    def forward(self, hidden_states: List[torch.Tensor], cls_tokens: List[torch.Tensor]) -> List[torch.Tensor]:
```
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/dpt/modeling_dpt.py:635]: Also breaking if it cannot be None

File: src/transformers/models/dpt/modeling_dpt.py [hunk 13]
```diff
@@ -1036,6 +1074,10 @@ def forward(self, hidden_states: List[torch.Tensor]) -> torch.Tensor:
         # use last features
         hidden_states = hidden_states[self.config.head_in_index]
 
+        if self.projection is not None:
+            hidden_states = self.projection(hidden_states)
+            hidden_states = ACT2FN["relu"](hidden_states)
```
Comment (ArthurZucker) [inline src/transformers/models/dpt/modeling_dpt.py:1079]:
```suggestion
            hidden_states = nn.ReLu(hidden_states)
```
No need to use ACT2FN here no?

**Review by ArthurZucker — COMMENTED** [review]
Thanks a mile for adding support to more models! 
✅ for the changes to the image processor 
✅ for the conversion scripts, if the tests are optional and in separate functions.
✅ for the changes to dino, if we make them BC
🟨 for the changes to DPT, that's a lot of changes. @amyeroberts did not seem against it, but wondering if it makes sense to have a new model? Are we stuck because DINOv2 uses it and we would have a lack of consistency?

File: src/transformers/models/dinov2/modeling_dinov2.py (hunk 9 above)
Comment (NielsRogge) [reply src/transformers/models/dinov2/modeling_dinov2.py:893]: Weird this wasn't catched by the `test_model_outputs_equivalence` test.. thanks for spotting!

Comment (NielsRogge) [reply src/transformers/models/dinov2/modeling_dinov2.py:894]: Fixed

File: src/transformers/models/dpt/image_processing_dpt.py (hunk 11 above)
Comment (NielsRogge) [reply src/transformers/models/dpt/image_processing_dpt.py:221]: No, so will close this comment

**Review by amyeroberts — COMMENTED** [review]
Thanks for iterating! 

Just a few small things to address before merging

File: src/transformers/models/dpt/convert_dinov2_depth_to_hf.py (hunk 10 above)
Comment (NielsRogge) [reply src/transformers/models/dpt/convert_dinov2_depth_to_hf.py:331]: I've added a `--verify_logits` flag

**Comment by NielsRogge** [comment]
@amyeroberts thanks for your review, I've addressed all comments. Feel free to merge :)

File: tests/models/dpt/test_modeling_dpt_auto_backbone.py [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+    """
+
+    all_model_classes = (DPTForDepthEstimation,) if is_torch_available() else ()
+    pipeline_model_mapping = (
+        {
+            "depth-estimation": DPTForDepthEstimation,
+        }
+        if is_torch_available()
+        else {}
+    )
```
Comment (amyeroberts) [inline tests/models/dpt/test_modeling_dpt_auto_backbone.py:146]:
```suggestion
    pipeline_model_mapping = {"depth-estimation": DPTForDepthEstimation} if is_torch_available() else {}
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for iterating! 

A few final small comments / things to address. 

General comment: I completely agree with @ArthurZucker's comments about the added complexity to the model. I'm not a fan of adding this if/else structure in the forward method. However, I don't see an easy way to change because of the `backbone_out_indices` argument and the weight loading already mapping to `self.dpt`.

**Comment by NielsRogge** [comment]
I've addressed all your comments, feel free to merge. To do:

- [ ] update all checkpoints on the hub to use `size_divisor` rather than `size_divisbility` in the preprocessor_config.json

**Review by amyeroberts — APPROVED** [review]
Thanks for iterating!

**Comment by amyeroberts** [comment]
@NielsRogge Thanks again for this addition - merged! I'll let you handle the updates to the configs on the hub.

**Comment by Starlento** [comment]
Hello guys. Thank you for adding this. I have a silly question that is there any plan that facebook will push this to the huggingface repo? Like `facebook/dpt-dinov2-large-nyu`? And could you can kindly inform me who is actually operating the companies accounts... Seems it is not all done by their side based on my guess.

**Comment by NielsRogge** [comment]
Hi @Starlento,

All models are on the hub: https://huggingface.co/models?pipeline_tag=depth-estimation&other=dinov2&sort=trending. I'll open a PR to make this more explicit in the docs

**Comment by lightonthefloor** [comment]
I want to know if it is possible to release the training part, and if it is possible, when will it be released?

File: src/transformers/models/dpt/modeling_dpt.py [hunk 15]
```diff
@@ -1109,32 +1150,46 @@ def forward(
...
+                    feature
+                    for idx, feature in enumerate(hidden_states[1:])
+                    if idx in self.config.backbone_out_indices[2:]
+                )
 
-            hidden_states = backbone_hidden_states
+                hidden_states = backbone_hidden_states
+
+        patch_height, patch_width = None, None
+        if self.config.backbone_config is not None and self.config.is_hybrid is False:
```
Comment (idol791) [inline src/transformers/models/dpt/modeling_dpt.py:1186]: Why calculate `patch_height` and `patch_width` (the input image size as multiples of patch_size) only under these conditions? If backbone_config is None, `self.config.patch_size` can be used instead of `self.config.backbone_config.patch_size`.
@NielsRogge
